### ISSUE TITLE
Wrong OpenStack + Cinder instance IDs

### ISSUE THREAD

**Issue by jsafrane** [kubernetes/kubernetes#13556]
I'm playing with Cinder volume plugin and I noticed that it expects that the machine hostname is the same as the OpenStack instance ID.

In our OpenStack deployment, the hostname is `<instance ID>.long.domain.redhat.com`. Kubernetes OpenStack cloud provider, especially `getComputeIDbyHostname()`, takes this long host name and tries to find an instance with this ID, which obviously fails and Cinder plugin does not mount any volumes.

I don't know if this error is caused just by our local setup or it could be common bug for many OpenStack deployments.

@spothanis, any opinion how it should work?

**Comment by spothanis** [comment]
@jsafrane The main idea is to find out on which node (esp. instance id) this volume operation is supposed to happen, for that I thought hostname was a decent choice and `getComputeIDbyHostname` tries to do that. It checks if the hostname reported by the machine matches server metadata reported by nova for fields -- hostname or name. 
In our installation of openstack, hostname reported by the machine matches nova's metadata field `hostname`
In your installation could you please check the metadata fields that are returned by nova for any server. If it matches any other fields I would be more than happy add them.
Also please let me know if you have any thoughts on how this can be done without the hostname matching :)

**Comment by jsafrane** [comment]
I am not OpenStack guy, I am just testing it. I do not know what's usual there. In my environment, server.Metadata is empty and hostname is probably the only way.

It would be nice if the code checked also value of `kubelet --hostname-override=<instanceID>` instead of `ossys.Hostname()`. Or a new element in cloud.conf?

**Comment by spothanis** [comment]
Just to make sure when you say in "your environment server.metadata is empty" you are referring to metadata returned by `nova show <instance-id>` command right?

**Comment by jsafrane** [comment]
Yes. `nova show` displays no metadata.
```console
$ nova show jsafrane-k2
+--------------------------------------+-----------------------------------------------------------------+
| Property                             | Value                                                           |
+--------------------------------------+-----------------------------------------------------------------+
| OS-DCF:diskConfig                    | MANUAL                                                          |
| OS-EXT-AZ:availability_zone          | nova                                                            |
| OS-EXT-STS:power_state               | 1                                                               |
| OS-EXT-STS:task_state                | -                                                               |
| OS-EXT-STS:vm_state                  | active                                                          |
| OS-SRV-USG:launched_at               | 2015-09-03T11:17:13.000000                                      |
| OS-SRV-USG:terminated_at             | -                                                               |
| accessIPv4                           |                                                                 |
| accessIPv6                           |                                                                 |
| config_drive                         |                                                                 |
| created                              | 2015-09-03T11:17:01Z                                            |
| flavor                               | m1.large (900e71fc-70cb-4913-8650-975c89c69ec9)                 |
| hostId                               | f9c8708aef9cd773c04a92b7df9142b323d22ab2bb54123c074a0a9f        |
| id                                   | dbdfceba-0023-46a7-9d76-1824da4dc538                            |
| image                                | Fedora 21 Cloud Official (cb324b06-7a4a-4bb9-a57a-ed6f407deafb) |
| key_name                             | jsafrane2                                                       |
| metadata                             | {}                                                              |
| name                                 | jsafrane-k2                                                     |
| os-extended-volumes:volumes_attached | []                                                              |
| os1-internal-1381 network            | <censored>                                        |
| progress                             | 0                                                               |
| security_groups                      | default                                                         |
| status                               | ACTIVE                                                          |
| tenant_id                            | 31c8c0e250b840ca94b2d96d3e83b5df                                |
| updated                              | 2015-09-03T11:17:13Z                                            |
| user_id                              | 92a80918755d435f95b0b79dd6dcec9e                                |
+--------------------------------------+-----------------------------------------------------------------+
```

And `OpenStack.getComputeIDbyHostname()` gets empty `server.Metadata` at openstack.go:912.

**Comment by spothanis** [comment]
Let me see what other options we have that can be used to derive the instance id. will get back to you soon.

**Comment by jsafrane** [comment]
cc @spothanis, what's your opinion?

**Comment by HackToday** [comment]
Hi @jsafrane @spothanis  I followed in guide 
https://github.com/kubernetes/kubernetes/tree/master/examples/mysql-cinder-pd

it can not work for kubectl create pods with such mysql yml file, the mysql container is created, but cinder volume was not attached or used by any pod(containers).


Did you try it successfully ?  I used k8s v1.1 official release

**Comment by spothanis** [comment]
@HackToday are you seeing any errors in kubelet logs?

**Comment by HackToday** [comment]
@spothanis  I used fedora VM, and with
journalctl -r -u kubelet, did not find any error logs.

I found pod is created, but  cinder volume was not attached.  


Below are details:

1.Create a cinder volume. Run "cinder create --display-name=myvoltest 1 "

  "cinder list"
   volumeID:0a9c53b7-97f2-4369-8d62-43a2885c6be8, status:available, Name:myvoltest, Size:1, Volume Type:lvmdriver-1, Bootable:false, Multiattach:False

2.Config the file "/etc/kubernetes/cloud_config""
 
 Content in "/etc/kubernete/cloud_config" file:
 
###
# kubernetes kubelet (minion) config

# The address for the info server to serve on (set to 0.0.0.0 or "" for all interfaces)
KUBELET_ADDRESS="--address=0.0.0.0"

# The port for the info server to serve on
# KUBELET_PORT="--port=10250"

# You may leave this blank to use the actual hostname
KUBELET_HOSTNAME=""

# location of the api-server
KUBELET_API_SERVER="--api_servers=http://*****:8080"

# Add your own!
KUBELET_ARGS="--cloud-provider=openstack --cloud-config=/etc/kubernetes/cloud_config"
[minion@k8-taoivuiqp5-0-dugpbsrwtsvz-kube-minion-q6lgoyatvozg kubernetes]$ ls
apiserver  cloud_config  config  controller-manager  examples  kubelet  kubelet.bak  proxy  scheduler
[minion@k8-taoivuiqp5-0-dugpbsrwtsvz-kube-minion-q6lgoyatvozg kubernetes]$ cat cloud_config
[Global]
auth-url=http://****:35357/v2.0
username=admin
password=password
region=RegionOne
tenant-id=****


3. Write the cloud_config to the "/etc/kubernete/kubelet"

Content in /etc/kubernete/kubelet
###
# kubernetes kubelet (minion) config

# The address for the info server to serve on (set to 0.0.0.0 or "" for all interfaces)
KUBELET_ADDRESS="--address=0.0.0.0"

# The port for the info server to serve on
# KUBELET_PORT="--port=10250"

# You may leave this blank to use the actual hostname
KUBELET_HOSTNAME=""

# location of the api-server
KUBELET_API_SERVER="--api_servers=http://192.168.0.3:8080"

# Add your own!
KUBELET_ARGS="--cloud-provider=openstack --cloud-config=/etc/kubernetes/cloud_config"


4. Run the command "cluster/kubectl.sh create -f examples/mysql-cinder-pd/mysql.yaml" in http://kubernetes.io/v1.1/examples/mysql-cinder-pd/README.html
   The result of command is error validating "/home/minion/kubernetes/examples/mysql-cinder-pd/mysql-service.yaml": error validating data: field nodePort: is required; if you choose to ignore these errors, turn validation off with --validate=false. 

5. Then run "cluster/kubectl.sh --validate=false create -f examples/mysql-cinder-pd/mysql.yaml" and "cluster/kubectl.sh --validate=false create -f examples/mysql-cinder-pd/mysql-service.yaml"
The result of command is that the cinder volume doesn't be mounted. 
If I ignore the error and add the  --validate=false like above. It doesn't have the error.
However, The cinder volume doesn't be count. run the command "sudo docker inspect ****":

"Mounts": [
{
"Source": "/var/lib/kubelet/pods/6f73be8b-8cfa-11e5-9380-fa163ec0b6bd/volumes/kubernetes.io~empty-dir/mysql-persistent-storage",
"Destination": "/var/lib/mysql",
"Mode": "",
"RW": true
},
{
"Source": "/var/lib/kubelet/pods/6f73be8b-8cfa-11e5-9380-fa163ec0b6bd/containers/mysql/9159778bf436d255493716da0ab754b547b64e012780edbea9ae21628a6e7f61",
"Destination": "/dev/termination-log",
"Mode": "",
"RW": true
}
]



Could you give some help about it ? Thanks

**Comment by jsafrane** [comment]
@HackToday, please run kubelet with `--v=5`, try to attach a volume and and post the logs somewhere, there must be some error there.

**Comment by HackToday** [comment]
@jsafrane and @spothanis seems you mentioned that

https://github.com/kubernetes/kubernetes/issues/17360

for kubernetes to work with openstack cinder volume plugin, it was added in kubernete 1.1. If that is true, it means our kubernetes version is old (1.04)

**Comment by tg90nor** [comment]
I also have a problem with hostname not matching instance name. In our setup openstack adds .novalocal to hostname. There is no metadata value that specifies the instance name only.

Maybe there should be a field in the openstack cloud config file, to specify hostname suffix, so it can be stripped away?

**Comment by anguslees** [comment]
@tg90nor: I think https://github.com/jsafrane/kubernetes/commit/7fc8bcae60b14cb59f803a53868172218fc6b034 addresses your issue.

**Comment by tg90nor** [comment]
Indeed it does. Thanks!

**Comment by saad-ali** [comment]
Any updates on this?

**Comment by anguslees** [comment]
@saad-ali My understanding is that it was fixed by https://github.com/kubernetes/kubernetes/pull/15537, merged in November.

I'd close this bug, but I don't think I'm allowed to do so...

**Comment by saad-ali** [comment]
Great, thanks! Closing as resolved by #15537
